README.md
# creditguard-node

> A simple wrapper for [creditguard](http://creditguard.co.il) api

[![styled with prettier](https://img.shields.io/badge/styled_with-prettier-ff69b4.svg)](https://github.com/prettier/prettier)

## Usage

### Installation

```
$ npm install creditguard
```

### Setting up

```js
var creditguard = require('creditguard');

var env = {
  user: 'username',
  password: 'password',
  server: 'https://cguat2.creditguard.co.il',
  terminal: '1234567',
  mid: '531', // required only for redirect
  // optional callback addresses when using cg redirect api
  // when left empty, cg will use the predefined terminal value
  success_url: 'localhost/payment/proceed?',
  error_url: 'localhost/payment/rejected?',
  cancel_url: 'localhost/payment/rejected?',
  // optional interface language of MPI hosted payment page
  language: 'heb' // default vaue is 'eng'
};

var options = {
  cleanup: true // remove empty fields from result, default to false
};

var cg = creditguard(env, options);
```

### Charging

`creditguard-node` follows cg xml-api (see their docs) parameters names. For example, if their api expects the following xml :

```xml
<ashrait>
   <request>
      <command>doDeal</command>
      <requestId/>
      <version>1001</version>
      <language>Eng</language>
      <mayBeDuplicate>0</mayBeDuplicate>
      <doDeal>
         <terminalNumber>0962XXX</terminalNumber>
         <cardNo>458045XXXXXX4580</cardNo>
         <cardExpiration>1212</cardExpiration>
         <creditType>RegularCredit</creditType>
         <currency>USD</currency>
         <transactionCode>Phone</transactionCode>
         <transactionType>Debit</transactionType>
         <total>10010</total>
         <validation>AutoComm</validation>
         <user>567890</user>
      </doDeal>
   </request>
</ashrait>
```

Then the proper use would be :

```js
// basically create an object with everything you want to put under 'doDeal' element
// (no need to inset terminal number or any other value from 'env')
let charge = {
  cardNo: '458045XXXXXX4580',
  cardExpiration: '1212'
  creditType: 'RegularCredit',
  currency: 'USD',
  transactionCode: 'Phone',
  transactionType: 'Debit',
  total: 10010,
  validation: 'AutoComm'
  user: '567890'
};

try {
  let res = await cg.call(charge)
  // ...
} catch (err) {
  // ...
}
```

### Customer data

If you are using the xml api to get a redirect page and you want to use `customerData` fields to pass some data into that page, you can do it simply by attaching it to `charge` object:

```js
charge.customerData = {
  userData1: 'userData1',
  userData3: 'userData3',
  ...
}
```

### Invoices

If your terminal supports invoices you can easily attach invoice to the call, again, by attaching it to the `charge` object:

```js
// check out cg invoice documentation for complete list of properties
charge.invoice = {
  invoiceCreationMethod: 'wait',
  invoiceSubject: 'Subject',
  invoiceItemQuantity: 1,
  invoiceItemPrice: 10010,
  companyInfo: 'Vandelay Industries',
  mailTo 'george@vandelay.com',
  ...
}
```

### Response

`creditguard-node` takes cg `xml` response and converts it to a `javascript` object so there's no need for you to parse any xml. It also verifies that the transaction result code is `000` (OK) and if not, throws an informative error that you can catch.

## TODO

* Add more tests.
* Add more apis for common operations like charge, capture and redirect.


CHANGELOG.md
## Changelog

* v1.0.0 - moved to async await and return `tranId` in response
* v0.0.5 - added an option to show MPI hosted payment page in user-defined language
* v0.0.4 - turned `mid` and `*_url` fields to optional to allow non-redirect api calls
* v0.0.3 - removed bad unicode characters that can jam the transaction
* v0.0.2 - added xml header for ISO-8859-8 encoding


test/test-charge.js
const expect = require('chai').expect;
const creditguard = require('../lib/creditguard');

const payment_env = {
  language: 'heb',
  user: 'israeli',
  password: 'PASSWORD',
  server: 'https://cguat2.creditguard.co.il',
  terminal: 'TERM'
};

const payment_env_with_mid = {
  language: 'heb',
  user: 'israeli',
  password: 'PASSWORD',
  server: 'https://cguat2.creditguard.co.il',
  terminal: 'TERM',
  mid: 'MID',
  success_url: 'http://localhost:3000/success',
  error_url: 'http://localhost:3000/error',
  cancel_url: 'http://localhost:3000/cancel'
};

const charge = {
  cardNo: '4012888888881881',
  cardExpiration: '0916',
  id: '310608195',
  cvv: '123',
  total: 100,
  transactionType: 'Debit',
  creditType: 'RegularCredit',
  currency: 'ILS',
  transactionCode: 'Phone',
  validation: 'AutoComm',
  uniqueid: '1234',
  email: 'test@yourown.com',
  invoice: {
    invoiceCreationMethod: 'wait',
    invoiceSubject: 'your invoice',
    invoiceItemCode: '1234',
    invoiceItemDescription: 'test',
    invoiceItemPrice: 100,
    invoiceItemQuantity: 1,
    isItemPriceWithTax: 1,
    companyInfo: 'yourown info',
    mailTo: 'test@yourown.com',
    ccDate: new Date().toISOString().slice(0, 10)
  }
};

describe('charge', async () => {
  it('should successfully charge', async done => {
    const cg = creditguard(payment_env);
    try {
      const res = await cg.call(charge);
      expect(res).to.exist;
      done();
    } catch (err) {
      done(err);
    }
  });

  it('should successfully charge with mid and urls', async done => {
    const cg = creditguard(payment_env_with_mid);
    try {
      const res = await cg.call(charge);
      expect(res).to.exist;
      done();
    } catch (err) {
      done(err);
    }
  });
});


test/lib/mock-server.js
var express = require('express');
var xml2js = require('xml2js');
var jsonxml = require('jsontoxml');

var parser_opts = {
  emptyTag: undefined,
  ignoreAttrs: true,
  explicitArray: false,
  explicitRoot: false,
};

var xmljson = new xml2js.Parser(parser_opts).parseString;

var env = {
  address: 'http://localhost:6000',
  terminal: '1234567',
  mid: '123',
  username: 'user',
  password: 'pass',
};

exports.get_server = function() {

  var app = express();

  app.use(express.bodyParser());

  app.post('/xpo/Relay', function(req, res) {
    // TODO need to emulate cg server
  });

  // start the mock cg server on port 6000
  app.port = 6000;
  app.listen(app.port);



  return app;
};

exports.address = env.address;
exports.terminal = env.terminal;
exports.mid = env.mid;
exports.username = env.username;
exports.password = env.password;



lib/creditguard.js
const xml2js = require('xml2js');
const jsonxml = require('jsontoxml');
const request = require('request-promise');
const pd = require('pretty-data').pd;
const debug = require('debug')('creditguard');

const parserOpts = {
  emptyTag: undefined,
  ignoreAttrs: true,
  explicitArray: false,
  explicitRoot: false
};

// an array of decimal values of unicode characters that create problems
const badChars = [8235, 8206, 8236, 1523];

const xmljson = new xml2js.Parser(parserOpts).parseString;

const defaults = {
  currency: 'ILS',
  creditType: 'RegularCredit',
  validation: 'AutoComm' // this means charge immediately
};

module.exports = function creditguard(env, options) {
  env = env || {};
  if (!env.user) throw new Error('env.user is required');
  if (!env.password) throw new Error('env.password is required');
  if (!env.server) throw new Error('env.server is required');
  if (!env.terminal) throw new Error('env.terminal is required');
  // Make the field mid optional for using as XML API without redirection
  if (env.mid) {
    env.success_url = env.success_url || '';
    env.error_url = env.error_url || '';
    env.cancel_url = env.cancel_url || '';
  }

  options = options || {};
  const cleanup = options.cleanup;

  const api = {};

  api.call = async function(options) {
    const doDeal = setDefaults(options, defaults);

    doDeal.terminalNumber = env.terminal;
    if (env.mid) {
      doDeal.mid = env.mid;
      doDeal.successUrl = env.success_url;
      doDeal.errorUrl = env.error_url;
      doDeal.cancelUrl = env.cancel_url;
    }

    const requestObj = {
      ashrait: {
        request: {
          version: '1001',
          language: env.language || 'eng',
          command: 'doDeal',
          doDeal: doDeal
        }
      }
    };

    return sendRequest(requestObj);
  };

  async function sendRequest(requestObj) {
    const header = "<?xml version='1.0' encoding='ISO-8859-8'?>";
    let requestXml = jsonxml(requestObj, {
      prettyPrint: true,
      indent: '    '
    });
    requestXml = stripCharacters(requestXml);
    requestXml = header.concat(requestXml);
    const req = {
      resolveWithFullResponse: true,
      url: env.server + '/xpo/Relay',
      method: 'post',
      form: {
        user: env.user,
        password: env.password,
        int_in: requestXml
      }
    };

    debug('------ sending xml request to crediguard ------>');
    debug(requestXml);
    debug('<-----------------------------------------------');

    const xmlResponse = await request(req);

    if (xmlResponse.statusCode !== 200) {
      throw new Error(`Unexpected response from gateway server`);
    }
    const { body: xmlBody } = xmlResponse;

    debug('------ response from creditguard ------>');
    debug(pd.xml(xmlBody));
    debug('<------------------------------------->');

    const json = await xmlToJSON(xmlBody);

    const { response: { result, doDeal, userMessage, tranId } } = json;

    if (result !== '000') {
      throw new Error(userMessage);
    }

    if (cleanup) {
      clean(result);
    }

    return Object.assign(doDeal, { tranId });
  }

  return api;
};

async function xmlToJSON(xml) {
  return new Promise((resolve, reject) => {
    xmljson(xml, function(err, result) {
      if (err) {
        return reject(err);
      }
      return resolve(result);
    });
  });
}

// recursively remove all empty properties from object
function clean(obj) {
  for (var property in obj) {
    if (obj.hasOwnProperty(property)) {
      if (typeof obj[property] === 'object') {
        clean(obj[property]);
      } else {
        if (!obj[property]) {
          delete obj[property];
        }
      }
    }
  }
}

function stripCharacters(str) {
  let exp = '';
  badChars.forEach(function(c) {
    exp += String.fromCharCode(c) + '|';
  });
  const regex = new RegExp(exp, 'g');
  return str.replace(regex, '');
}

function setDefaults(obj) {
  const length = arguments.length;
  if (length < 2 || obj == null) {
    return obj;
  }
  for (let index = 1; index < length; index++) {
    const source = arguments[index],
      keys = Object.keys(source),
      l = keys.length;
    for (let i = 0; i < l; i++) {
      const key = keys[i];
      if (obj[key] === void 0) {
        obj[key] = source[key];
      }
    }
  }
  return obj;
}
